### ISSUE TITLE
[REQ] Allow overriding `__init__` of struct.dataclass

### ISSUE THREAD

**Issue by PhilipVinc** [google/flax#1292]
Python data classes allow the ovverriding of frozen data classes `__init__` methods, that essentially just set all the relevant fields correctly.

I would like to ask that you expose that kwargument in your data class wrapper.

This is useful if a frozen data class is usually constructed starting from object_A but does not store object_A itself, but rather something computed starting from it.
We can get around this by using a constructor method, but for a series of reasons the easiest things for us to do in NetKet would be to just override the `__init__`.

I don't see reasons not to allow this, as it is allowed by the standard dataclass specification.

**Comment by PhilipVinc** [comment]
To give a more concrete example, we have a series of LinearOperator-like objects called `FMJacobian` and `FMOnTheFly` and `FMJacobianPyTree` (for example) which are lazy encodings of the fisher matrix (~hessian/natural gradient). 

From an high level API they all take as input a model, the training data and the parameters, to be constructed.
However `FMJacobian` computes the `jax.grad(model, pars, data)` and ravels it to a dense array at construction and stores it. `FMJacobianPyTree` stores the gradient. `FMOnTheFly` stores model/data/parameters and precomputes nothing...

Ideally we'd do this in the `__init__` method.
Again, we _can_ use a factory method, but we would rather not because, among over reasons...

 - We need to remember in type annotations to annotate with `FMJacobianT` but construct it with `FMJacobian`. 
 - we use annotations together with (multiple)dispatch to change behaviour. Having different names is bug prone and annoying.

**Comment by jheek** [comment]
This question pops up quite often but there is a major issue. A smart `__init__` method (that is one where args don't map 1:1 to attributes) typically don't support clone & replace properly.

So in practice:
For a functional dataclass we want to update fields by calling cls(**dataclass_fields_as_dict(instance), **updates). This is only well defined if __init__ simply maps the constructor args to attributes.

**Comment by PhilipVinc** [comment]
Ah, makes sense. Though it remains annoying.

I guess you need to flatten/unflatten the data class, right?
Because an alternative would be for you to add a custom flatten/unflatten_init function.

**Comment by PhilipVinc** [comment]
cc @femtobit @attila-i-szabo with whom we were discussing this

**Comment by jheek** [comment]
You can always define a vanilla dataclass with a custom flatten/unflatten. But in this case it's your own responsibility to reconstruct without a trivial `constructor(*fields)` call.

Disclaimer I'm a big Rust fanboy:
But I think "smart" constructors are not a good pattern. I think a constructor should always be simple and what you want instead is a classmethod that creates your dataclass from some other value.
A simple argument for this is that once the constructor starts being smart you could imagine wanting multiple constructors. But to maintain sanity these constructors should probably have a self-explanatory name. 
So I think what you should do is something like: `FMJacobian.from_model(model, params, data)`

**Comment by marcvanzee** [comment]
@jheek Would it make sense to move this to a GH Discussion?
